File: tests/core/matrices/codac2_tests_GaussJordan.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2025
 *  \author     Damien Massé
 *  \copyright  Copyright 2025 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_GaussJordan.h>
#include <codac2_BoolInterval.h>
#include <codac2_IntervalVector.h>
#include <iostream>

using namespace std;
using namespace codac2;

TEST_CASE("GaussJordan")
{
  /* rectangular matrix, full rank */
  {
    Matrix M 
    { { 1, -4, 6, 7, 6 },
      { 2,  1, 3, 6, -2 },
      { 5,  2, 2, 9, -1 } };
    Matrix pre = gauss_jordan(M);
    /* check that pre*M is "band matrix" */
    Matrix P = pre*M;
    double mx = 0.0;
    for (Index i=1;i<M.rows();i++) 
      for (Index j=0;j<i;j++) {
        mx = std::max(mx,std::abs(P(i,j)));
        mx = std::max(mx,std::abs(P(M.rows()-1-i,M.cols()-1-j)));
    }
    CHECK(mx<1e-10);
  }
  
}

File: tests/core/matrices/codac2_tests_GaussJordan.py
#!/usr/bin/env python
 
#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2025
#  \author     Damien Massé
#  \copyright  Copyright 2025 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestGaussJordan(unittest.TestCase):
   
    def test_GaussJordan(self):

       M = Matrix([
            [ 1, -4, 6, 7, 6 ],
            [ 2,  1, 3, 6, -2 ],
            [ 5,  2, 2, 9, -1 ] ])
       pre = gauss_jordan(M)
       # check that pre*M is "band matrix" */
       P = pre*M
       mx = 0.0
       for i in range(1,M.rows()-1):
          for j in range(0,i-1):
             mx = max(mx,abs(P(i,j)))
             mx = max(mx,abs(P(M.rows()-1-i,M.cols()-1-j)))
       self.assertTrue(mx<1e-10)


if __name__ ==  '__main__':
  unittest.main()

